core/src/main/java/io/questdb/cairo/wal/WalWriter.java

```diff
@@ -147,23 +147,22 @@ public long apply(AlterOperation operation, boolean contextAllowsAnyStructureCha
         if (operation.isStructureChange()) {
             long txn;
             do {
+                boolean structureVersionMismatch = false;
                 try {
                     alterOperationValidationBackend.startAlterValidation();
                     operation.apply(alterOperationValidationBackend, true);
-                    if (alterOperationValidationBackend.structureVersion != metadata.getStructureVersion() + 1) {
-                        throw CairoException.nonCritical().put("table structure change did not contain 1 transaction [table=").put(tableName)
-                                .put(", oldStructureVersion=").put(metadata.getStructureVersion())
-                                .put(", newStructureVersion=").put(alterOperationValidationBackend.structureVersion).put(']');
-                    }
-                } catch (SqlException e) {
+                    structureVersionMismatch = (alterOperationValidationBackend.structureVersion != metadata.getStructureVersion() + 1);
+                } catch (CairoException e) {
                     // Table schema (metadata) changed and this Alter is not valid anymore.
                     // Try to update WAL metadata to latest and repeat one more time.
                     goActive();
-                    try {
-                        operation.apply(alterOperationValidationBackend, true);
-                    } catch (SqlException e2) {
-                        throw CairoException.nonCritical().put(e2.getFlyweightMessage());
-                    }
+                    operation.apply(alterOperationValidationBackend, true);
+                }
+
+                if (structureVersionMismatch) {
+                    throw CairoException.nonCritical().put("table structure change does not contain 1 transaction [table=").put(tableName)
```

Comment:
this message should read:

```
could not apply transaction, there is a gap in structure version numbers
```

Additionally it should be `critical` and perhaps should also suggest how to recover from this situation.

Reply:
Changed in f2fdf7294dd3c08b887a904694d74e1a849d5c58